lib/charms/grafana_agent_k8s/v0/grafana_agent.py

```diff
@@ -155,37 +192,88 @@ def _on_log_proxy_relation_departed(self, event):
         self._update_agents_list(event)
 
         if len(self._current_config["clients"]) == 0:
-            self._container.stop(self._container_name)
+            self._container.stop("promtail")
         else:
-            self._container.restart(self._container_name)
+            self._container.restart("promtail")
 
     def _on_upgrade_charm(self, event):
         # TODO: Implement it ;-)
         pass
 
-    def _on_promtail_pebble_ready(self, _):
-        """Event handler for the pebble ready event.
+    def _get_container(self, container_name):
```

Comment:
I changed this implementation (not yet committed).

For instance in the consumer side you can send the container_name

```python
self._log_proxy = LogProxyConsumer(charm=self, log_files=LOG_FILES, container_name="workload_app")
```
And in the lib:

```python
    def __init__(
        self,
        charm,
        log_files: list,
        container_name: Optional[str],
        relation_name: str = DEFAULT_RELATION_NAME,
    ):
    ...

    def _get_container(self, container_name):
        if container_name is not None:
            return self._charm.unit.get_container(container_name)

        containers = dict(self._charm.model.unit.containers)

        if len(containers) == 1:
            return self._charm.unit.get_container([*containers].pop())

        # FIXME: Use custom exception
        raise Exception("Container cannot be obtained")
```

The idea is: 
- If you send the container_name, we'll use that name to get the container.
- If you do not send the container_name:
  - If there is only one container in the Pod, we'll use it
  - If there is more than one container we'll raise an Exception
  
  What do you think about this?

Reply:
That is what I was looking for, except the Exception message that provides no guidance whatsoever to understand what is wrong with it :-)

Go with:

```
No 'container_name' parameter has been specified; since this charmed operator is not running exactly one container, it must be specified which container to get logs from.
```

Note: there will be soon charms WITHOUT workload containers so the error message must account for the "0 containers" case.